docs/pyspark-pandas-udf.md

```diff
@@ -0,0 +1,40 @@
+Pyspark Pandas Udf
+==================
+split-apply-merge is a useful pattern when analyzing data. It is implemented in many popular data analying libraries such as Spark, Pandas, R, and etc. Split and merge operations in these libraries are similar to each other, mostly implemented by a `group by` operator. For instance, Spark DataFrame has groupBy, Pandas DataFrame has groupby. Therefore, for users familiar with either Spark DataFrame or pandas DataFrame, it is not difficult for them to understand how grouping works in the other library. However, `apply` is more native to different libraries and therefore, quite different between libraries. A pandas user knows how to use `apply` to do a linear regression on the grouped data might not know how to do the same calculation using pyspark, for instance. Also, the current implementation of passing data from the java executor to python executor is not effiecient, there is oppurtunity to speed it up using Apache Arrow. We are proposing new functions that allows easy and high performance split-apply-merge computation using pyspark and pandas.
+
+Related Work
+============
+SPARK-13534
+
+This enables faster data serialization between Pyspark and Pandas using Apache Arrow. Our work will be on top of this and use the same serialization for pandas udf.
+
+SPARK-12919 and SPARK-12922
+
+These implemented two functions: dapply and gapply in Spark R which implements the similar split-apply-merge pattern that we want to implement with Pyspark.
+
+API
+===
+We propose adding a pandas udf function:
```

Comment:
I think there are three questions a user needs to answer in order for the engine to understand the intended execution, and these will drive the design of the API:

1. Which subsets of the data do we want to see passed to the UDF?
   a. Spark partitions (maybe overlapped partitions in Flint)
   b. Disjoint subsets defined by partition key (`df.groupby('key').apply(fn)`)
   c. Overlapping subsets defined by sort order and window size, optionally also divided into subsets by partition key (`df.groupby('key').window('5 days').apply(fn)`)
2. What are the input and output types and array shapes?  In R, the function name encodes this (`aaply`, `ddply`, `dlply`, etc.).
   In pandas, there are `transform` (Transformation: perform some group-specific computations and return a like-indexed.), `aggregate`/`sum`/etc. (Aggregation: computing a summary statistic (or statistics) about each group.), and `apply` (can do a lot, including filtering).
   In pandas, `transform` maps each of the columns in each group through the function, which should map a `Series` to a same-length `Series`, `aggregate` maps each of the columns in each group through the function, which should reduce a `Series` to a scalar value, and `apply` does some guesswork based on calling your function with the whole group (all columns of it) as a `DataFrame`, and seeing what you give back (it then infers what the resulting thing should look like, based on whether you return a scalar, a `Series`, or another `DataFrame`).
   Here are the function signatures I think make sense to support (and I'd like to give them separate names, and avoid how pandas tries to guess in `apply`):
   a. `Series -> scalar`: Aggregate each column of each group individually to a scalar, the result is a grouped structure with one row per group, which contains the result of the aggregation.
   b. `DataFrame -> Row (or tuple)`: Aggregate the whole `DataFrame` to a single row, the result is a grouped structure with one row per group, which contains the result of the aggregation.
   c. `Series -> Series`: Transform each column of each group separately to a new column, presumably this would append the resulting column to the Spark `DataFrame` like `withColumn`.
   d. `DataFrame -> Series`: Probably more useful, transform maybe several columns of the `DataFrame` to produce a new column, again, we would probably append this column to the Spark `DataFrame`.
   e. `DataFrame -> DataFrame`: Possibly we would compute several new columns, or modify existing ones (not a fan of this, obviously) all in one call to the python interpreter.
   Note that only _a_ and _b_ are probably candidates for application to windows or overlapping subsets, since it's unclear what you do with return values from different calls in which the overlapped rows participated.
3. What is the schema after the operation, probably.

I think two good next steps for this document are:
1. Write down a motivating example of a computation we want to support this way (e.g. rolling average?)
2. Take my text above and try to incorporate it, in a way that expresses the matrix of possible APIs we could offer.  Feel free to copy my words or write your own.  In addition to this, it might be good to, for each of these APIs, write out what the user documentation would look like for it, and write down an example or two of how you would use it (what would the UDF look like)